File: 02_Stack/04_Daily_Temperatures/739_Daily_Temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] res = new int[temperatures.length];
        Stack<Integer> s = new Stack<>();

        for (int i=0; i < temperatures.length; i++){
            while(!s.empty() && temperatures[i] > temperatures[s.peek()]){
                int old_day = s.peek();
                s.pop();
                res[old_day] = i - old_day;
            }
            s.push(i);
        }

        return res;
    }
}

File: 02_Stack/04_Daily_Temperatures/739_Daily_Temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> res(temperatures.size(), 0);
        stack<int> s;

        for (int i = 0; i < temperatures.size(); i++){
            cout << i;

            while (!s.empty() && temperatures[i] > temperatures[s.top()]){
                cout << "ok";
                int old_day = s.top();
                s.pop();
                res[old_day] = i - old_day; 
            }
            s.push(i);
        }

        return res;
    }
};
